### PR TITLE
WIP : Added the first version of scil_make_priors_functionnectome

### PR DESCRIPTION
**Pull request by NotaCS**
First version of the Pythonized script to create the anatomical priors.
Pull request suggested by Arnaud.

### CODE REVIEW

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by NotaCS** [comment]
Ideally, this script should be able to run without actually installing the whole scilpy library (because I want it to be incorporated with the Functionnectome, which will be installed by non-developers, who might mess-up their python environment?

File: scripts/scil_make_priors_functionnectome.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+    trk_list = glob.glob(os.path.join(args.in_dir_tractogram, '*.trk'))
+    for trk_F in trk_list:
+        print(os.path.basename(trk_F))
+        trk_sft = load_tractogram(trk_F, templ_i, bbox_valid_check=False)
+        _ = trk_sft.remove_invalid_streamlines()  # Ideally, only remove the sections outsite of the boundingbox
+        trk_sft = filter_streamlines_by_length(trk_sft, 25, 250)
+        trk_sft.to_vox()
+        for vox in vox_list:  # TODO Parallelize
+            vox_mask = base_mask.copy()
```
Comment (frheault) [inline scripts/scil_make_priors_functionnectome.py:68]: I think the next few lines could be replaced by
filter_grid_roi(sft, mask, filter_type, is_exclude): in segment/streamlines.py of scilpy

File: scripts/scil_make_priors_functionnectome.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            vox_mask[tuple(vox)] = True
+
+            vox_strm = trk_sft.get_streamlines_copy()
+            vox_strm = Streamlines(utils.target(vox_strm, templ_i.affine, vox_mask))
+            vox_strm = vox_strm.get_streamlines_copy()
+
+            vol_strm = np.zeros(templ_i.shape)
```
Comment (frheault) [inline scripts/scil_make_priors_functionnectome.py:75]: The next few lines should be replaced by compute_tract_counts_map() in tractanalysis/streamlines_metrics

File: scripts/scil_make_priors_functionnectome.py (hunk 1 above)
Comment (NotaCS) [reply scripts/scil_make_priors_functionnectome.py:68]: Hi François, thanks for pointing out that function, I think it's exactly what I need! It can even filter end-points, which is perfect!

**Comment by NotaCS** [comment]
Is there a function to crop the streamlines that have a portion outside of the bounding-box? For now, it's completely removing the said voxels, but I just want to remove the bad sections (typically, it's sections that are in the lower brain-stem / upper spinal cord).

File: scripts/scil_make_priors_functionnectome.py (hunk 1 above)
Comment (NotaCS) [reply scripts/scil_make_priors_functionnectome.py:68]: C'est étrange, `target` et `filter_grid_roi` ne me donnent pas le même nombre de streamlines avec un même tractogram et mask (filter_grid_roi en donne plus). Est-ce que filter_grid_roi est plus liberal dans sa définition de ce qui est dans le mask ?

Comment (frheault) [reply scripts/scil_make_priors_functionnectome.py:68]: Je crois que la difference est dû à un shift, habituellement il faut un sft.to_corner() (pour définir ce qu'est le centre d'un voxel). Tu ne le faisais pas, mais la fonction le fait. Peut-être que c'est ça.

File: scripts/scil_make_priors_functionnectome.py (hunk 2 above)
Comment (frheault) [reply scripts/scil_make_priors_functionnectome.py:75]: This function computes a density map (that you can then binarize). But it is robust to large step size. Doing it by hand will skip a few voxel if you step sizes are above 0.5mm, while the cython function does a line integration.

image

**Comment by frheault** [comment]
@NotaCS from scilpy.utils.streamlines import cut_invalid_streamlines

That will cut the streamlines that are out. It is somewhat slow, but as good as it gets in python.

File: scripts/scil_make_priors_functionnectome.py (hunk 2 above)
Comment (NotaCS) [reply scripts/scil_make_priors_functionnectome.py:75]: OK, I just tested it, no problem. Though it dosn't give the exact same voxels as my script... so I think I'll have to check how it count the streamlines

**Comment by NotaCS** [comment]
> @NotaCS from scilpy.utils.streamlines import cut_invalid_streamlines
> 
> That will cut the streamlines that are out. It is somewhat slow, but as good as it gets in python.

@frheault Thank you! Indeed, it's quite slow. But in the end, with the other functions you gave me, it seems that we don't even need it anymore. the streamlines sections outside of the bounding-box will just be ignored when creating the final map.

File: scripts/scil_make_priors_functionnectome.py (hunk 2 above)
Comment (NotaCS) [reply scripts/scil_make_priors_functionnectome.py:75]: The tractogram I'm using has a step size of 0.5mm, but voxels of 2mm in the final map, so it shouldn't be a problem. It might be due to the shift you mentioned in the other comment. I'll investigate, thanks!

**Comment by NotaCS** [comment]
Alright, so actually, I still needed to remove the bad streamline points... So I made my own function!
It's rougher but much much faster than cut_invalid_streamlines (6 sec instead of 6 min).

**Comment by frheault** [comment]
@NotaCS The code is cryptic, what is the difference?

sft.remove_invalid_streamlines() remove the entire streamline if it contains a single invalid point.
cut_invalid_streamlines() keeps the longest possible segment that is made only of valid streamlines.

What is yours doing? From what I understand it just crops out invalid points and reconnects the streamlines after, which is not a behavior I trust fully (But I doubt that in human data it's a problem, so it's fair).

I saw the test 'strm[0] > epsilon'. This is a check for the lower corner of the grid. You also need a test for the upper corner of the grid: Which is strm[0] < grid.shape-epsilon (It is a 3D test since the shape is not always the same value in x/y/z)

Also that test seems to remove the streamlines that have their first point and last out of the bounding box. Which is the vast majority of case (in the CST or top of the CC) It is very rare to enter and exit back a volume. At this point you should simply use sft.remove_invalid_streamlines()? No?

We can talk on Teams if thats faster.

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by NotaCS** [comment]
New, full version available, with a parallel processing option and region-wise priors creation.
For now, it is just way too slow, as filter_grid_roi takes 8 sec to generate filter the whole tractogram, and we have to use it on all the voxels (about 230,000 times), for every subjet (100), so that's not possible.
I'm working on a function to reduce the size of the tractogram to speed this up.

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by NotaCS** [comment]
With the new version, one the map for one voxel (only for one subject) takes now less than 2sec. With parallelisation of the process (say, no 32 gpu), it should take: (time per voxel) * (nb of voxel) * (nb of subjects) / (nb of cpu) = 1.5*228,336*100/32 = 1070325 sec = 300h = about 12 days....
Unless I'm missing something, it's still way too long (but it's much better than before).
Any suggestion?

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by NotaCS** [comment]
I just pushed the new code, much faster, but still a bit too slow for my test (it would take 3-4 days on 32cpu to run it on the 100 subjects). I'm trying to implement another approach proposed by Guillaume that may speed things up quite a bit

**Comment by NotaCS** [comment]
New version, iterating over streamlines instead of voxels (in a first step at least), as per Guillaume advice. Much (much!) faster. But gives a "Broken pipe" error if I run it on too many (>=6) CPUs at the same time on my local machine.

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by NotaCS** [comment]
Mmmmhh... It doesn't work well. I don't know why, but it only spawned 3 workers (when I asked 6), one got stuck quickly, and my RAM usage went through the roof (although each processes should only have to work with about 3 or 4 GB of data, max)...

File: scripts/scil_make_priors_functionnectome.py [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+            if nb_proc == 1:
+                print('Starting process')
+                strm_multi(vox_dict, stream_tupled, args.out_dir, templ_i)
+            else:  # Parallelize
+                vox_batchs = np.array_split(list(vox_dict), nb_proc)
+                vox_dict_batchs = [{tuple(v): vox_dict[tuple(v)] for v in vb} for vb in vox_batchs]
+                print('Starting parallel process')
+                with Pool(processes=nb_proc) as pool:
+                    poolCheck = pool.starmap_async(strm_multi, zip(vox_dict_batchs,
+                                                                   [stream_tupled]*nb_proc,
```
Comment (GuillaumeTh) [inline scripts/scil_make_priors_functionnectome.py:216]: Here instead of duplicate n times the data, prepare your batches before and give an array of batches with 1 batch per process. In the end you will just have 1 time the data in RAM instead of n times.

Comment (NotaCS) [reply scripts/scil_make_priors_functionnectome.py:216]: normally, [stream_tupled]*nb_proc does not make copy of the obejct (each item of the list points to the same original object). And I can't split stream_tupled into batches, because each process may need any streamline in it. Or am I missing something?

**Comment by NotaCS** [comment]
Alright, so it was indeed probably the starmap that was messing things up. Now it works perfectly. I could optimise the RAM a bit more (as each worker has a copy of the data right now), but it will do for now, it's not that hungry.

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by NotaCS** [comment]
Ran well: on a local machine, with 8 CPUs, it took a bit less than 17 hours and about 13GB of RAM. Perfect :)

**Comment by NotaCS** [comment]
Mmmh... I compared a map with one from the original priors. It's quite similar, but really not exactly the same. A quick correlation between the two (on the brain voxels only) gives me R = 0.9, which is quite low considering that they are supposed to be the same. Some voxels have a difference of up to 0.66 (knowing that a map ranges from 0 to 1, it's a huge discrepancy).

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by NotaCS** [comment]
When I put a tractogram in vox space, and align it to the corner, what exactly mean the coordinates? 
For example, does a streamline point at (15, 25, 14) means that I am looking at a point in the center of voxel (15, 25, 14), or at one of its corner?

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-497/11/

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-497/13/

**Comment by NotaCS** [comment]
I'll close the PR for now, to stop the automatic messages that must be sent to everyone each time I update something ;)
